txnsync/peer.go

```diff
@@ -228,26 +228,35 @@ func (p *Peer) selectPendingTransactions(pendingTransactions []transactions.Sign
 			continue
 		}
 
-		p.lastTransactionSelectionGroupCounter = pendingTransactions[grpIdx].GroupCounter
-
 		if windowSizedReached {
 			hasMorePendingTransactions = true
 			break
 		}
 		selectedTxns = append(selectedTxns, pendingTransactions[grpIdx])
 		selectedTxnIDs = append(selectedTxnIDs, txID)
 
-		// calculate the total size of the transaction group.
-		for txidx := range pendingTransactions[grpIdx].Transactions {
-			encodingBuf := protocol.GetEncodingBuf()
-			accumulatedSize += len(pendingTransactions[grpIdx].Transactions[txidx].MarshalMsg(encodingBuf))
-			protocol.PutEncodingBuf(encodingBuf)
-		}
+		// add the size of the transaction group
+		accumulatedSize += pendingTransactions[grpIdx].EncodedLength
+
 		if accumulatedSize > windowLengthBytes {
 			windowSizedReached = true
 		}
 	}
 
+	// update the lastTransactionSelectionGroupCounter if needed -
+	// if we selectged any transaction to be sent, update the lastTransactionSelectionGroupCounter with the latest
+	// group counter. If the startIndex was *after* the last pending transaction, it means that we don't
+	// need to update the lastTransactionSelectionGroupCounter since it's already ahead of everything in the pending transactions.
+	if grpIdx >= 0 && startIndex < len(pendingTransactions) {
+		if grpIdx == len(pendingTransactions) {
+			if grpIdx > 0 {
+				p.lastTransactionSelectionGroupCounter = pendingTransactions[grpIdx-1].GroupCounter + 1
+			}
+		} else {
+			p.lastTransactionSelectionGroupCounter = pendingTransactions[grpIdx].GroupCounter
```

Comment:
No `+ 1` here because the counter indicates the next group to be processed?

Reply:
exactly.